xapian-core/weight/tfidfweight.cc

```diff
@@ -49,10 +50,31 @@ TfIdfWeight::TfIdfWeight(const std::string &normals)
     need_stat(WDF_MAX);
 }
 
+TfIdfWeight::TfIdfWeight(const std::string &normals, double slope, double delta)
+    : normalizations(normals), param_slope(slope), param_delta(delta)
+{
+    if (normalizations.length() != 3 ||
+	!strchr("nbslP", normalizations[0]) ||
+	!strchr("ntpfsP", normalizations[1]) ||
+	!strchr("nP", normalizations[2]))
+	throw Xapian::InvalidArgumentError("Normalization string is invalid");
+    if (normalizations[1] != 'n') {
+	need_stat(TERMFREQ);
+	need_stat(COLLECTION_SIZE);
+    }
+    need_stat(WDF);
+    need_stat(WDF_MAX);
+    if (param_slope != 0 || param_delta != 0) {
+	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(WQF);
+    }
```

Comment:
That's not a good assumption - people may make the normalisations configurable, and then it's probably simpler to just call the most general constructor and always supply slope and delta values, rather than conditionalise which constructor to call based on the contents of the normalisation string.

Fundamentally, the checks for which stats we request should exactly match the conditions under which we use those stats - that's more robust and we don't have to worry about what happens in the cases where one condition is met and the other isn't.  We check for `P` in the normalisation string before we use them, so we need to check for that here too.